### PR TITLE
support like escape

### PR DESCRIPTION
**Pull request by siddontang**
Fix https://github.com/pingcap/tidb/issues/344

### CODE REVIEW

File: expression/like.go [hunk 1]
```diff
@@ -129,18 +135,23 @@ func compilePattern(pattern string) (patChars, patTypes []byte) {
...
+					// mysql will treat escape character as the origin value even
+					// the escape sequence is invalid in Go or C.
+					// e.g, \m is invalid in Go, but in MySQL we will get "m" for select '\m'.
+					// Following case is correct just for escape \, not for others like +.
+					// TODO: add more checks for other escapes.
 					i--
-					c = '\\'
+					c = escape
 				}
 			}
```
Comment (shenli) [inline expression/like.go]: How about i==len(pattern)-1 ?

Comment (siddontang) [inline expression/like.go]: I just use the old code here, which just seems using the origin escape character directly. 
I have added a comment that we will add more tests later.

File: parser/parser.y [hunk 2]
```diff
@@ -1510,9 +1514,17 @@ Factor1:
...
-		$$ = &expression.PatternLike{Expr: $1.(expression.Expression), Pattern: $4.(expression.Expression), Not: $2.(bool)}
+		escape := $5.(string)
+		if len(escape) > 1 {
+			yylex.(*lexer).errf("Incorrect arguments %s to ESCAPE", escape)
+			return 1
+		} else if len(escape) == 0 {
+			escape = "\\"
+		}
+		$$ = &expression.PatternLike{Expr: $1.(expression.Expression), Pattern: $4.(expression.Expression), 
+					Not: $2.(bool), Escape: escape[0]}
```
Comment (qiuyesuifeng) [inline parser/parser.y]: Alignment？

**Comment by qiuyesuifeng** [comment]
LGTM

File: expression/like.go (hunk 1 above)
Comment (coocood) [inline expression/like.go]: If the escape is the last character in the pattern, the match type would be exact match.

**Comment by coocood** [comment]
LGTM

**Pull request merged**